Fix flaky tests

```
-test.shuffle 1695385443914646887
--- FAIL: Test_get_attachment_RunE (0.01s)
    --- FAIL: Test_get_attachment_RunE/success (0.00s)
        get_attachment_test.go:81: STDOUT: To regenerate the on-disk snapshot, re-run the tests with `UPDATE_TESTY=true` or `-update-testy` or `-tags=update`.
            --- expected
            +++ actual
            @@ -1 +1 @@
            -Testing
            +
        get_attachment_test.go:81: STDERR: To regenerate the on-disk snapshot, re-run the tests with `UPDATE_TESTY=true` or `-update-testy` or `-tags=update`.
            --- expected
            +++ actual
            @@ -1 +1 @@
            -
            +Error: http: invalid Read on closed Body
        get_attachment_test.go:81: Unexpected exit status. Want 0, got 2
```

--- a/cmd/kivik/output/output_test.go
+++ b/cmd/kivik/output/output_test.go
@@ -17,6 +17,8 @@ import (
 	"io"
 	"strings"
 	"testing"
+
+	"github.com/google/go-cmp/cmp"
 )
 
 func Test_ensurenewlineEnding(t *testing.T) {
@@ -40,6 +42,33 @@ func Test_ensurenewlineEnding(t *testing.T) {
 		}
 		if buf.String() != "asdf\n" {
 			t.Errorf("Unexpected output: %q", buf.String())
+		}
+	})
+}
+
+func TestJSONReader(t *testing.T) {
+	t.Run("success", func(t *testing.T) {
+		in := map[string]interface{}{
+			"foo": "bar",
+		}
+		r := JSONReader(in)
+		got, err := io.ReadAll(r)
+		if err != nil {
+			t.Fatal(err)
+		}
+		want := `{"foo":"bar"}
+`
+		if d := cmp.Diff(want, string(got)); d != "" {
+			t.Error(d)
+		}
+	})
+	t.Run("error", func(t *testing.T) {
+		in := make(chan int)
+		r := JSONReader(in)
+		_, err := io.ReadAll(r)
+		want := "json: unsupported type: chan int"
+		if err == nil || err.Error() != want {
+			t.Errorf("Unexpected error: %s", err)
 		}
 	})
 }

--- a/cmd/kivik/output/util.go
+++ b/cmd/kivik/output/util.go
@@ -15,17 +15,37 @@ package output
 import (
 	"encoding/json"
 	"io"
+	"sync"
 	"text/template"
 )
 
+// delayedJSONReader waits until the first call to Read to call json.Marshal.
+// This saves calling json.Marshal if the reader is never read, and also
+// prevents the consumption of the input, in cases like an attachment, which
+// indirectly marshals an io.Reader.
+type delayedJSONReader struct {
+	mu sync.Mutex
+	r  io.Reader
+	i  interface{}
+}
+
+func (d *delayedJSONReader) Read(p []byte) (int, error) {
+	d.mu.Lock()
+	if d.r == nil {
+		r, w := io.Pipe()
+		go func() {
+			err := json.NewEncoder(w).Encode(d.i)
+			_ = w.CloseWithError(err)
+		}()
+		d.r = r
+	}
+	d.mu.Unlock()
+	return d.r.Read(p)
+}
+
 // JSONReader marshals i as JSON.
 func JSONReader(i interface{}) io.Reader {
-	r, w := io.Pipe()
-	go func() {
-		err := json.NewEncoder(w).Encode(i)
-		_ = w.CloseWithError(err)
-	}()
-	return r
+	return &delayedJSONReader{i: i}
 }
 
 // FriendlyOutput produces friendly output.

